Removed ClipboardMonitor from ClipboardAction.paste type button

Maksim-Nikolaev opened the pull request:
## Description

According the issue [#2421](https://github.com/singerdmx/flutter-quill/issues/2421) ClipboardMonitor was causing lag if "paste" button was enabled in QuillSimpleToolbar. 

That was because of the periodic 1 second check for Clipboard content. In case with large strings or images in clipboard, it would lag out the application and slow down everything drastically. 
```dart
class ClipboardMonitor {
  bool _canPaste = false;
  bool get canPaste => _canPaste;
  Timer? _timer;

  void monitorClipboard(bool add, void Function() listener) {
    if (kIsWeb) return;
    if (add) {
      _timer = Timer.periodic(
          const Duration(seconds: 1), (timer) => _update(listener));
    } else {
      _timer?.cancel();
    }
  }

  Future<void> _update(void Function() listener) async {
    final clipboardService = ClipboardServiceProvider.instance;
    if (await clipboardService.hasClipboardContent) {
      _canPaste = true;
      listener();
    }
  }
}
```

This is not a breaking change, however it will change the behavior of  availability for "paste" button for QuillSimpleToolbar since it always be enabled (except of readonly mode).

## Related Issues

- *Fix #2421*

## Type of Change

- [x] 🛠️ **Bug fix:** Resolves an issue without changing current behavior.

EchoEllet commented:
Thanks for your contribution.

> This is not a breaking change

A breaking change is not the kind that requires users to update their code to compile, but to also keep the current state without any behavior breaking.

Changing the default of a value is also a breaking change.

What would happen if the user pressed the paste button and there was nothing to paste?

I suggest providing the option to override the `canPaste` state instead, so they don't need to depend on `ClipboardMonitor`.

I'm with the idea of disabling something that's causing issues. However, users should have the option to revert to the old behavior in case they still want `ClipboardMonitor` with a signle boolean parameter.

Maksim-Nikolaev commented:
> A breaking change is not the kind that requires users to update their code to compile, but to also keep the current state without any behavior breaking. Changing the default of a value is also a breaking change.

Ok I understand

> What would happen if the user pressed the paste button and there was nothing to paste?

Nothing will happen. There are few checks that prevent breaking everything on calling "paste" function. 

I double checked it with Web & Android version and I strongly believe that iOS will not break either (don't have it right now, so I can't check)

QuillController.clipboardPaste handles this quite well, here is the full function:


```dart
  /// Returns whether paste operation was handled here.
  /// [updateEditor] is called if paste operation was successful.
  @experimental
  Future<bool> clipboardPaste({void Function()? updateEditor}) async {
    if (readOnly || !selection.isValid) return true;

    final clipboardConfig = config.clipboardConfig;

    if (await clipboardConfig?.onClipboardPaste?.call() == true) {
      updateEditor?.call();
      return true;
    }

    final pasteInternalImageSuccess = await _pasteInternalImage();
    if (pasteInternalImageSuccess) {
      updateEditor?.call();
      return true;
    }

    const enableExternalRichPasteDefault = true;
    if (clipboardConfig?.enableExternalRichPaste ??
        enableExternalRichPasteDefault) {
      final pasteHtmlSuccess = await pasteHTML();
      if (pasteHtmlSuccess) {
        updateEditor?.call();
        return true;
      }

      final pasteMarkdownSuccess = await pasteMarkdown();
      if (pasteMarkdownSuccess) {
        updateEditor?.call();
        return true;
      }
    }

    final clipboardService = ClipboardServiceProvider.instance;

    final onImagePaste = clipboardConfig?.onImagePaste;
    if (onImagePaste != null) {
      final imageBytes = await clipboardService.getImageFile();

      if (imageBytes != null) {
        final imageUrl = await onImagePaste(imageBytes);
        if (imageUrl != null) {
          replaceText(
            plainTextEditingValue.selection.end,
            0,
            BlockEmbed.image(imageUrl),
            null,
          );
          updateEditor?.call();
          return true;
        }
      }
    }

    final onGifPaste = clipboardConfig?.onGifPaste;
    if (onGifPaste != null) {
      final gifBytes = await clipboardService.getGifFile();
      if (gifBytes != null) {
        final gifUrl = await onGifPaste(gifBytes);
        if (gifUrl != null) {
          replaceText(
            plainTextEditingValue.selection.end,
            0,
            BlockEmbed.image(gifUrl),
            null,
          );
          updateEditor?.call();
          return true;
        }
      }
    }

    // Only process plain text if no image/gif was pasted.
    // Snapshot the input before using `await`.
    // See https://github.com/flutter/flutter/issues/11427
    final plainText = (await Clipboard.getData(Clipboard.kTextPlain))?.text;

    if (plainText != null) {
      final plainTextToPaste = await getTextToPaste(plainText);
      if (pastePlainTextOrDelta(plainTextToPaste,
          pastePlainText: _pastePlainText, pasteDelta: _pasteDelta)) {
        updateEditor?.call();
        return true;
      }
    }

    final onUnprocessedPaste = clipboardConfig?.onUnprocessedPaste;
    if (onUnprocessedPaste != null) {
      if (await onUnprocessedPaste()) {
        updateEditor?.call();
        return true;
      }
    }

    return false;
  }
```



> I suggest providing the option to override the `canPaste` state instead, so they don't need to depend on `ClipboardMonitor`.

I will then provide such option.
As an extra step, if we are keeping ClipboardMonitor, I suggest we optimize it at least to reduce the amount of checks. Can I make the timer to 5 seconds and prevent overlapping checks? (So it won't start checking if it didn't finish the previous one)

Maksim-Nikolaev commented:
Please check the most recent commit.
I added **bool? enableClipboardPaste** in lib\src\toolbar\config\simple_toolbar_config.dart for QuillSimpleToolbarConfig 

This is passed directly to ClipboardButton with .paste action

As for the ClipboardButton itself and monitor:

- [x] enableClipboardPaste flag can be null or bool:
```dart
  /// If null, uses in-built [ClipboardMonitor]
  /// If true, paste button is enabled (if true & not readonly)
  /// If false, paste button is disabled
  final bool? enableClipboardPaste;
```
- [x] Updated ClipboardMonitor periodic timer from 1 second to 5 seconds
- [x] Update ClipboardMonitor to not overlap with previous task of itself (in case of big clipboard content)
- [x] Changed getter for "currentStateValue" to prioritize the enableClipboardPaste flag:
```dart
        return !controller.readOnly &&
            (kIsWeb || (widget.enableClipboardPaste ?? _monitor.canPaste));
```

EchoEllet commented:
> Updated ClipboardMonitor periodic timer from 1 second to 5 seconds

This is still a breaking change. 
Could you minimize the changes to only override the can paste state (whether the button is enabled)?

> I added bool? enableClipboardPaste in lib\src\toolbar\config\simple_toolbar_config.dart for QuillSimpleToolbarConfig

This option should be only in the button config, not the toolbar config.

Maksim-Nikolaev commented:
> This is still a breaking change.

Reverted to 1 second

Maksim-Nikolaev commented:
> This option should be only in the button config, not the toolbar config.

I agree that it would be the best way to do it. Unfortunately I couldn't find the proper place to add a new value, could you elaborate here?
```dart
  final QuillToolbarToggleStyleButtonOptions clipboardPaste;
```

```dart
import 'package:flutter/foundation.dart' show immutable;
import 'package:meta/meta.dart';

import '../base_button_options.dart';

class QuillToolbarToggleStyleButtonExtraOptions
    extends QuillToolbarBaseButtonExtraOptions
    implements QuillToolbarBaseButtonExtraOptionsIsToggled {
  const QuillToolbarToggleStyleButtonExtraOptions({
    required super.controller,
    required super.context,
    required super.onPressed,
    required this.isToggled,
  });

  @override
  final bool isToggled;
}

@immutable
class QuillToolbarToggleStyleButtonOptions
    extends QuillToolbarBaseButtonOptions<QuillToolbarToggleStyleButtonOptions,
        QuillToolbarToggleStyleButtonExtraOptions> {
  const QuillToolbarToggleStyleButtonOptions({
    super.iconData,
    super.iconSize,
    super.iconButtonFactor,
    super.tooltip,
    super.afterButtonPressed,
    super.iconTheme,
    super.childBuilder,
  });
}
```

Maksim-Nikolaev commented:
> Could you minimize the changes to only override the can paste state (whether the button is enabled)?

I would love to, but since it's coded in the way that we have little control over ClipboardMonitor it's difficult to update it's state or to stop the timer when the button already exists.

To stop the timer we need to trigger "didUpdateWidget" and add/remove extra listeners but it only happens when controller changes.
I kept the original handler to imitate the same behavior, as well as added some custom code to add/remove listener if "enableClipboardPaste" value changed from parent.

Maksim-Nikolaev commented:
I think adding "isEnabled" to **QuillToolbarBaseButtonOptions** would be the best option, but this would require adding the functionality of this option to all the buttons. I can do it if you give me a green light.
Otherwise I'm open for recommendations how to handle it for just 1 type of button

Maksim-Nikolaev commented:
- [x] Moved **enableClipboardPaste** to **QuillToolbarBaseButtonOptions**

On the video you can see that it works for all 3 cases (null, true, false) and the timer will continue as normal if it is set back to "null" 

https://github.com/user-attachments/assets/b076a992-27c8-4a51-8a5f-a6d38d5ea0f6

singerdmx commented:
@EchoEllet can we merge this?

EchoEllet commented:
> Updated ClipboardMonitor periodic timer from 1 second to 5 seconds

This is still a breaking change. 
Could you minimize the changes to only override the can paste state (whether the button is enabled)?

> I added bool? enableClipboardPaste in lib\src\toolbar\config\simple_toolbar_config.dart for QuillSimpleToolbarConfig

This option should be only in the button config, not the toolbar config.
 
> can we merge this?

I will work on it soon. However, it's not a high-priorty issue given that I have disabled the buttons by default (revert to the default behavior).

Maksim-Nikolaev commented:
> This is still a breaking change.

I adjusted the code changes followed by your requests. Check some comments above.
Timer is reverted back to 1 second
The flag is now in the button config 



I really appreciate your time and work on this!

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -16,27 +16,49 @@ class ClipboardMonitor {
+    // Prevent multiple clipboard checks at the same time.
+    _isCheckingClipboard = true;
+
+    // Initialize the clipboard service.
```
This is static, and a new instance will not be created. Could you remove comments for code that's already obvious?

We're trying to not over comment, and only comment when there is something confusing, not clear or the code is not self explanatory and descriptive, in other words anything that can't be understood from other developers perspective.

EchoEllet commented on lib/src/toolbar/config/base_button_options.dart:
```diff
@@ -65,6 +66,11 @@ class QuillToolbarBaseButtonOptions<T, I> {
   /// Use custom theme
   final QuillIconTheme? iconTheme;
 
+  /// Disables the [ClipboardMonitor] of [QuillToolbarClipboardButton] if not null
+  /// If true or false, it will enable or disable the clipboard paste button
+  /// Only affects the clipboard paste button
+  final bool? enableClipboardPaste;
```
I suggest:

```dart
/// Determines if the paste button is enabled. 
///
/// Defaults to [ClipboardMonitor] in case of `null`, which will check if the clipboard has content to paste every second, and only then enable the button, indicating to the user that they can paste something.
///
/// Set to `true` to always enable it even if the clipboard has no content to paste, which will do nothing on a press.
```

However, it's not final and could use improvements. I will review it soon.

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -16,27 +16,49 @@ class ClipboardMonitor {
+    // Initialize the clipboard service.
     final clipboardService = ClipboardServiceProvider.instance;
+
+    // Check if the clipboard has content.
     if (await clipboardService.hasClipboardContent) {
+      // If the clipboard has content, set the flag to true.
```
It's already obvious what the code does, and over commenting can complicate the code readability.

EchoEllet commented on CHANGELOG.md:
```diff
@@ -10,6 +10,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Added **enableClipboardPaste** flag to ButtonOptions to control clipboard paste button enabled state and ClipboardMonitor [#2421](https://github.com/singerdmx/flutter-quill/issues/2421)
+    - null (default): Original ClipboardMonitor behavior (1 second periodic checks)
+    - false: Disable ClipboardMonitor & button is disabled
+    - true: Disable ClipboardMonitor & button is enabled
+
+### Changed
+
+- Prevented ClipboardMonitor from doing multiple simultaneous clipboard checks to improve performance. [#2421](https://github.com/singerdmx/flutter-quill/issues/2421)
```
The quality of the `CHANGELOG.MD` is a priority and will need a few changes, I will work on it soon.

EchoEllet commented:
> I really appreciate your time and work on this!

Thanks, I appreciate your contributions, too.
There are a few changes, and I will work on it today.

Maksim-Nikolaev commented:
I applied some changes as you requested. Obviously feel free to change the wording/comments to your liking! Thanks again

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -16,27 +16,43 @@ class ClipboardMonitor {
 
 class QuillToolbarClipboardButton extends QuillToolbarToggleStyleBaseButton {
-  QuillToolbarClipboardButton({
+  const QuillToolbarClipboardButton({
```
There is almost no use in adding the `const`, the `controller` can never be const so neither `QuillToolbarClipboardButton` can. It would only make the commit history more complex, the code more complex, and slightly confusing and introducing them removing it (if needed in case of the constructor body) would be a breaking change. I was addicted to using the `const` keyword but it doesn't bring much.

![image](https://github.com/user-attachments/assets/e6c232c4-e959-4fbd-8505-d8908e2c0b8c)

Maksim-Nikolaev replied:
I see, my bad for not going in-depth with that. In my experience it's the default-ish way in Flutter to mark class constructors with const and even flutter analysis options usually prefers const there

EchoEllet replied:
> my bad

No worries, I just wanted to ensure that you're aware of this in case you're not.

> mark class constructors with const

You only need to mark a class, but not the subclasses of that class. I'm uncertain if the analysis has a lint for checking the `const` or `@immutable`.

> flutter analysis options usually prefers const there

That's in case it's possible to instate a constructor that can be `const`, or if the superclass marked with `@immutable` such as `Widget`, the `StatelessWidget` and `StatefulWidget` doesn't need `@immutable` annotation.

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -66,23 +82,57 @@ class QuillToolbarClipboardButtonState
-      widget._monitor.monitorClipboard(true, _listenClipboardStatus);
+    if (widget.clipboardAction == ClipboardAction.paste &&
+        widget.options.enableClipboardPaste == null) {
+      _monitor.monitorClipboard(true, _listenClipboardStatus);
     }
   }
 
   @override
   void removeExtraListener(covariant QuillToolbarClipboardButton oldWidget) {
-    if (widget.clipboardAction == ClipboardAction.paste) {
-      oldWidget._monitor.monitorClipboard(false, _listenClipboardStatus);
+    if (widget.clipboardAction == ClipboardAction.paste &&
+        widget.options.enableClipboardPaste == null) {
+      _monitor.monitorClipboard(false, _listenClipboardStatus);
```
Could you add a private getter called `shouldUseClipboardMonitor` that is only `true` when `enableClipboardPaste` is `null` and then update related parts such as those to make use of this to make it easier and faster to understand from other developers's perspective.

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -100,6 +147,11 @@ class QuillToolbarClipboardButtonState
         ClipboardAction.paste => Icons.paste_outlined,
       };
 
+  bool get shouldUseClipboardMonitor {
```
Unless there is a need, it's preferable to keep this private (`_shouldUseClipboardMonitor` instead).

EchoEllet commented on lib/src/toolbar/config/base_button_options.dart:
```diff
@@ -65,6 +66,13 @@ class QuillToolbarBaseButtonOptions<T, I> {
   /// Use custom theme
   final QuillIconTheme? iconTheme;
 
+  /// Determines if the paste button is enabled. The button is disabled and cannot be clicked if set to `false`.
+  ///
+  /// Defaults to [ClipboardMonitor] in case of `null`, which checks if the clipboard has content to paste every second and only then enables the button, indicating to the user that they can paste something.
+  ///
+  /// Set it to `true` to enable it even if the clipboard has no content to paste, which will do nothing on a press.
+  final bool? enableClipboardPaste;
+
```
* I'm confused, why is this change in the `QuillToolbarBaseButtonOptions`? This is the base for all buttons, and any configuration specific to the paste button should be in `QuillToolbarPasteButton` instead.
* It's necessary to document if this is only applicable if the clipboard action is `ClipboardAction.paste`.

Maksim-Nikolaev replied:
Could you please take care of the location where to put the "enableClipboardPaste" flag? 

ClipboardButtons use "QuillToolbarToggleStyleButtonOptions" and there is no "QuillToolbarPasteButton" class. 
I'm not 100% familiar with how to introduce it without changes to be breaking.

```dart
  final QuillToolbarToggleStyleButtonOptions clipboardCut;
  final QuillToolbarToggleStyleButtonOptions clipboardCopy;
  final QuillToolbarToggleStyleButtonOptions clipboardPaste;
```

EchoEllet replied:
ClipboardButtons use "QuillToolbarToggleStyleButtonOptions".

It looks like the `QuillToolbarClipboardButton` doesn't have a custom config class (`QuillToolbarClipboardButtonConfig`), I'm not sure why but it shouldn't even if it doesn't require extra parameters. I will create a config class for this button and you can move this flag there instead.

> here is no "QuillToolbarPasteButton" class.

I updated my message, I meant the config class of `QuillToolbarClipboardButton` which is missing.

EchoEllet replied:
> ntroduce it without changes to be breaking.

You're right, it's not possible without any breaking in case users were already using `QuillToolbarToggleStyleButtonOptions` for those buttons (see #2433).

To be honest I'm not motivated to work on anything related to `QuillSimpleToolbar`, the design is a pain to work with and almost always confusing despite it's not very complex, our long plan is to completely rework the toolbar with a new modern one, better API design and usability while keeping the old toolbar backward compatible.

If #2433 merged, you can access `enableClipboardPaste` inside `QuillToolbarClipboardButtonState`:

```dart
widget._options.enableClipboardPaste;
```

But again, wouldn't it be better to mark this feature as experimental, rework on it later (to avoid breaking)? It's almost not a priority at least IMO. Working on it is not a good time investment which is why I updated `README.md` to include a note about contributing: https://github.com/singerdmx/flutter-quill#-contributing

@CatHood0 Do you think it worth introducing this breaking (changing `options` from `QuillToolbarToggleStyleButtonOptions` to `QuillToolbarClipboardButtonOptions` and slightly inconsistent API).

Maksim-Nikolaev replied:
> But again, wouldn't it be better to mark this feature as experimental, rework on it later (to avoid breaking)? It's almost not a priority at least IMO. Working on it is not a good time investment

I can understand the pain and now I realize more how hard it is to maintain open source project. I discovered that the paste button caused **extreme** lag in the app I'm working on, therefore I thought I'd contribute.

Otherwise in my case the usage of such button can be simplified to calling "controller.clipboardPaste" and can be achieved with custom button. 

Once again, really appreciate your time and patience with this PR and I hope I don't cause too much issues for you

EchoEllet replied:
> I don't cause too much issues for you

This is a project issue and not exclusive to this PR.

>  I realize more how hard it is to maintain open source project.

It's not very hard if you have a good solid foundation from the start, but there was a time when we kept merging PRs without much review and also introducing changes without much consideration or any tests at all, and that made it harder for us to fix those issues or maintain the project.

For example, the magnifier feature was introduced and merged quickly that people started depending on it (see #2406 for a list of regressions and issues), which caused major issues that affected the editor experience, and reverting to the old behavior is no longer trivial (see #2413), introducing breaking changes caused many issues for the app developers. That's why I think adding new features is not a priority, especially those that add more on the surface for more issues, this is not an issue with your code.

EchoEllet replied:
I have merged #2433, please rebase and use:

```dart
widget._options.enableClipboardPaste;
```

While also removing those two lines:

```diff
- // TODO: This field will be used by the PR: https://github.com/singerdmx/flutter-quill/pull/2427
- // ignore: unused_field
final QuillToolbarClipboardButtonOptions? _options;
```

> [!NOTE]
> This is far from an efficient solution, but we will continue with it as we plan on reworking the toolbar at a later point.

Maksim-Nikolaev replied:
I applied the changes, please review 👍

EchoEllet replied:
> It's necessary to document if this is only applicable if the clipboard action is ClipboardAction.paste. Since the QuillToolbarClipboardButton is the same widget for cut, copy, and paste buttons.

This is still not addressed.

Maksim-Nikolaev replied:
Removed it now from QuillToolbarBaseButtonOptions

EchoEllet commented on lib/src/toolbar/buttons/clipboard_button.dart:
```diff
@@ -143,16 +193,17 @@ class QuillToolbarClipboardButtonState
     }
 
     return UtilityWidgets.maybeTooltip(
-        message: tooltip,
-        child: QuillToolbarIconButton(
-          icon: Icon(
-            iconData,
-            size: iconSize * iconButtonFactor,
-          ),
-          isSelected: false,
-          onPressed: currentValue ? _onPressed : null,
-          afterPressed: afterButtonPressed,
-          iconTheme: iconTheme,
-        ));
+      message: tooltip,
+      child: QuillToolbarIconButton(
+        icon: Icon(
+          iconData,
+          size: iconSize * iconButtonFactor,
+        ),
+        isSelected: false,
+        onPressed: currentValue ? _onPressed : null,
+        afterPressed: afterButtonPressed,
+        iconTheme: iconTheme,
+      ),
+    );
```
If the trailing comma (`,`) is the only change, then please revert, this is a common style inconsistency and it should be fixed in a separate PR (#2442).

This change will make it harder to track the commit history, mainly we want to be able to see the PR and the commit that made a feature.

EchoEllet replied:
Fixed.

EchoEllet commented:
>  info • Constructors in '@immutable' classes should be declared as 'const' • lib/src/toolbar/buttons/clipboard_button.dart:62:3 • prefer_const_constructors_in_immutables
   info • Field overrides a field inherited from 'QuillToolbarBaseButtonOptions' • lib/src/toolbar/config/buttons/clipboard_button_options.dart:28:15 • overridden_fields
   info • The member 'enableClipboardPaste' overrides an inherited member but isn't annotated with '@override' • lib/src/toolbar/config/buttons/clipboard_button_options.dart:28:15 • annotate_overrides

Could you confirm if you have rebased your branch?

```shell
$ git remote add upstream git@github.com:singerdmx/flutter-quill.git
$ git fetch upstream
$ git rebase upstream/master # Fix conflicts if any. Follow the git CLI instructions.
# git push --force # Force push if conflicts were solved or didn't occur.
```

EchoEllet commented on lib/src/toolbar/config/buttons/toggle_style_options.dart:
```diff
@@ -29,5 +29,6 @@ class QuillToolbarToggleStyleButtonOptions
     super.afterButtonPressed,
     super.iconTheme,
     super.childBuilder,
+    super.enableClipboardPaste,
```
Could you revert any changes made to [`toggle_style_options.dart`](https://github.com/singerdmx/flutter-quill/pull/2427/files#diff-9258316c22a048d566a87172e61c3e053607f1dd38273c8fbdfcd57ff3ad06a7)?

EchoEllet commented:
> Error: Failed to create outline comment: Resource not accessible by integration

Could you [rebase your branch](https://github.com/singerdmx/flutter-quill/pull/2427#issuecomment-2598621102) once again?

EchoEllet commented:
> Error: Failed to create outline comment: Resource not accessible by integration

Could you [rebase your branch](https://github.com/singerdmx/flutter-quill/pull/2427#issuecomment-2598621102) once again?

Maksim-Nikolaev commented:
> Could you [rebase your branch](https://github.com/singerdmx/flutter-quill/pull/2427#issuecomment-2598621102) once again?

Rebased again.

> Please fix the analysis issue. https://dart.dev/tools/linter-rules/prefer_const_constructors_in_immutables

Added const to QuillToolbarClipboardButton constructor.
